orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,550 @@
+      gathergrad_input_args,
+      gathergrad_output_args,
+      {},
+      kMSDomain,
+      logger);
+
+  new_gathergrad_node->SetExecutionProviderType(node.GetExecutionProviderType());
+  auto gathergrad_out_arg = new_gathergrad_node->MutableOutputDefs()[0];
+
+  InlinedVector<NodeArg*> reshape_input_args;
+  reshape_input_args.push_back(gathergrad_out_arg);
+  reshape_input_args.push_back(unflattened_shape_arg);
+  InlinedVector<NodeArg*> reshape_output_args{&graph.GetOrCreateNodeArg(graph.GenerateNodeArgName("reshape_result"),
+                                                                        nullptr)};
+  Node* new_reshape_node = onnxruntime::optimizer::compute_optimizer::InsertIntermediateNodeOnDestInput(
+      graph, node,
+      in_index,
+      0,
+      0,
+      graph.GenerateNodeName("RecoverShape"),
+      "Reshape",
+      "Reshape node to recover invalid tokens.",
+      reshape_input_args,
+      reshape_output_args,
+      {},
+      kOnnxDomain,
+      logger);
+  new_reshape_node->SetExecutionProviderType(node.GetExecutionProviderType());
+  return new_reshape_node->MutableOutputDefs()[0];
+}
+
+Status PaddingElimination::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  LOG_DEBUG_INFO(logger, "Enter PaddingElimination");
+
+  if (sparse_embedding_input_names_.size() == 0) {
+    LOG_DEBUG_INFO(logger, "Exit PaddingElimination, no sparse embedding input names.");
+    return Status::OK();
+  }
+
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  Node* embedding_node = nullptr;
+  NodeArg* input_ids_arg = nullptr;
+  // Make sure each egde in candidate_edges has two consecutive dims to be flattened.
+  // Now only support the case that the first two dims are consecutive.
+  // All edges in candidate_edges constitute the subgraph which is propagated from the embedding node
+  std::unordered_set<NodeArg*> candidate_edges;
+  // nodes in candidate_input are the input nodes to the subgraph constituted by candidate_edges
+  // input edges of nodes in candidate_input should be in candidate_edges or to be added Reshape + Gather
+  std::unordered_set<Node*> candidate_input;
+  // nodes in candidate_output are the output nodes from the subgraph constituted by candidate_edges
+  // input edges of nodes in candidate_output, if in candidate_edges, should be added GatherGrad + Reshape
+  std::unordered_set<Node*> candidate_output;
+  std::queue<Node*> to_visit;
+  std::unordered_set<Node*> visited;
+
+  // Find the valid embedding node
+  for (auto node_index : node_topology_list) {
+    auto& node = *graph.GetNode(node_index);
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (IsATenEmbedding(&node) &&
+        graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders()) &&
+        node.InputDefs().size() >= 3 &&
+        node.InputDefs()[2]->Exists() &&
```

Comment:
node.InputDefs()[2] is used for few times, let's declare a variable with a meaninful name for example "padding_idx_input_arg"

Reply:
node.InputDefs()[1] and node.InputDefs()[2] is only used in this if statement. It's not convenient to declare a variable if not dividing the if statement.